subprojects/gradle/process/src/main/kotlin/com/avito/utils/ProcessRunner.kt

```diff
@@ -1,129 +1,137 @@
+            workingDirectory: File? = null,
+            outputTo: File? = null
+        ): Process {
+            return ProcessBuilder(*splitCommand(command))
+                .redirectErrorStream(true)
+                .also { builder ->
+                    if (workingDirectory != null) {
+                        builder.directory(workingDirectory)
+                    }
+                    if (outputTo != null) {
+                        builder.redirectOutput(outputTo)
+                    }
+                }
+                .start()
+        }
 
-    val output = processOutput.toString().trim()
+        private class Timeout(val time: Long, val unit: TimeUnit)
+
+        private fun runCommand(
+            command: String,
+            workingDirectory: File? = null,
+            timeout: Timeout?
+        ): Result<String> = try {
+            val process = spawnProcess(
+                command = command,
+                workingDirectory = workingDirectory
+            )
+            val outputFuture = cliOutputReaderThreadPool.submit(
+                CLIOutputReaderTask(
+                    stdout = process.inputStream
+                )
+            )
+            val output = if (timeout != null) {
+                process.waitFor(timeout.time, timeout.unit)
+                outputFuture.get(timeout.time, timeout.unit)
+            } else {
+                process.waitFor()
+                outputFuture.get()
+            }
 
-    if (process.exitValue() != 0) {
-        Result.Failure(Throwable("Unknown error: exit code=[${process.exitValue()}]; output=$output"))
-    } else {
-        Result.Success(output)
+            if (process.exitValue() != 0) {
+                Result.Failure(Throwable("Unknown error: exit code=[${process.exitValue()}]; output=$output"))
+            } else {
+                Result.Success(output)
+            }
+        } catch (t: Throwable) {
+            Result.Failure(t)
+        }
     }
-} catch (t: Throwable) {
-    Result.Failure(t)
 }
 
-private class IOThread(
-    val source: InputStream,
-    val result: StringBuffer,
-    val logger: Logger
-) : Thread() {
-
-    init {
-        isDaemon = true
-    }
-
-    override fun run() {
-        val reader = BufferedReader(InputStreamReader(source))
-
-        var line: String? = reader.readLine()
-        while (line != null) {
-            result.appendLine(line)
-
-            logger.debug(line)
-
-            line = reader.readLine()
+private class CLIOutputReaderTask(
+    private val stdout: InputStream
+) : Callable<String> {
+
+    override fun call(): String {
+        val result = StringBuilder()
+        val stdoutBuffer = BufferedReader(InputStreamReader(stdout))
+        while (true) {
+            val line = stdoutBuffer.readLine()
+            if (line != null) {
+                result.appendLine(line)
+            } else {
+                break
+            }
         }
+        return result.toString().trim()
```

Comment:
Why do we trim output?
What if spaces are meaningful? We can't know it on this level.

Reply:
I don't know but trim works only for the first and last space, so it's not critical but I could remove this logic